packages/frontend/src/features/sqlRunner/store/thunks.ts

```diff
@@ -0,0 +1,181 @@
+import {
+    scheduleSqlJob,
+    type ResultsAndColumns,
+} from '../hooks/useSqlQueryRun';
+import { SqlRunnerResultsRunnerFrontend } from '../runners/SqlRunnerResultsRunnerFrontend';
+import { selectSqlRunnerResultsRunner } from './sqlRunnerSlice';
+
+/**
+ * Run a sql query and return the results
+ * @param sql - The sql query to run
+ * @param limit - The limit of results to return
+ * @param projectUuid - The project uuid to run the query on
+ * @returns The results and the results runner
+ */
+export const runSqlQuery = createAsyncThunk<
+    ResultsAndColumns & {
+        resultsRunner: SqlRunnerResultsRunnerFrontend;
+        fileUrl: string | undefined;
+    },
+    { sql: string; limit: number; projectUuid: string },
+    { rejectValue: ApiErrorDetail }
+>(
+    'sqlRunner/runSqlQuery',
+    async ({ sql, limit, projectUuid }, { rejectWithValue }) => {
+        try {
+            const scheduledJob = await scheduleSqlJob({
+                projectUuid,
+                sql,
+                limit,
+            });
+
+            const job = await getSqlRunnerCompleteJob(scheduledJob.jobId);
+            if (isApiSqlRunnerJobSuccessResponse(job)) {
+                const url =
+                    job.details && !isErrorDetails(job.details)
+                        ? job.details.fileUrl
+                        : undefined;
+
+                const results = await getResultsFromStream<RawResultRow>(url);
+
+                const columns =
+                    job.details && !isErrorDetails(job.details)
+                        ? job.details.columns
+                        : [];
+
+                const resultsRunner = new SqlRunnerResultsRunnerFrontend({
+                    columns,
+                    rows: results,
+                    projectUuid,
+                    limit,
+                    sql,
+                });
+
+                return {
+                    fileUrl: url,
+                    results,
+                    columns,
+                    resultsRunner,
+                };
+            } else {
+                return rejectWithValue(job.error);
+            }
+        } catch (error) {
+            return rejectWithValue(error as ApiErrorDetail);
+        }
+    },
+);
+
+/**
+ * Fetch pivot chart data
+ * @param vizDataModel - The viz data model to fetch the pivot chart data for
+ * @param limit - The limit of results to return
+ * @param sql - The sql query to run
+ * @returns The pivot chart data
+ */
+export const fetchPivotChartData = createAsyncThunk<
+    PivotChartData | undefined,
+    {
+        vizDataModel:
+            | TableDataModel
+            | PieChartDataModel
+            | CartesianChartDataModel;
+        limit: number;
+        sql: string;
+    }
+>(
+    'cartesianChartBaseConfig/fetchPivotChartData',
+    async ({ vizDataModel, limit, sql }) => {
+        const chartData = await vizDataModel.getPivotedChartData({
```

Comment:
within this function there is secretly api calls happening using the react-query query client directly (i.e. without hooks). We might want to pull those out eventually

Reply:
Oh I had not seen this - it really goes deep in - we can use something different if you want.